external: update healthchecker caps for rbd command

subhamkrai opened the pull request:
<!-- Please take a look at our Contributing documentation before submitting a Pull Request!
https://rook.io/docs/rook/latest/Contributing/development-flow/

Thank you for contributing to Rook! -->

**Description of your changes:**
when creating networkFence CR, it requires IP's to block which we get from running `rbd status ...` command. But, the client.healthchecker user didn't had the right caps to run hence it was giving error. Now, adding the required caps `profile rbd-read-only` to osd so that rbd command can be executed.

**Which issue is resolved by this Pull Request:**
Resolves #12931 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

subhamkrai commented:
testing result, created  a new user test just for testing 
```
# In toolbox see the user caps
# ceph auth ls | grep -E 'test|healthchecker' -A 4
installed auth entries:

client.healthchecker
	key: AQCUcmNk/N2GDxAA6KMjeCBMNxFMtzxOUiMwOQ==
	caps: [mgr] allow command config
	caps: [mon] allow r, allow command quorum_status, allow command version
	caps: [osd] allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index
--
client.test
	key: AQB1Xw1lR4xwIRAATN9P2hdm6SHWh4hWcSxLFA==
	caps: [mgr] allow command config
	caps: [mon] allow r, allow command quorum_status, allow command version
	caps: [osd] profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index
```
```
# In operator pod, running the rbd command
srai@fedora ~ $  oc exec -it rook-ceph-operator-5d94564c97-wfqgt -- sh
sh-5.1$  rbd status rbd/csi-vol-22e6259d-90ff-460f-9adc-c401995d0e8b --cluster=openshift-storage --conf=/var/lib/rook/openshift-storage/openshift-storage.config --keyring=/var/lib/rook/openshift-storage/client.test --name=client.test
Watchers:
	watcher=10.1.160.189:0/2353215614 client.4404720 cookie=18446462598732840967
sh-5.1$ 
```
```

Madhu-1 commented:
@subhamkrai are you able to test the feature completely with this user? i mean even the `ceph osd blocklist` was successful?

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -632,7 +632,8 @@ def __init__(self, arg_list=None):
             "mgr": "allow command config",
             "mon": "allow r, allow command quorum_status, allow command version",
-            "osd": "allow rwx pool={0}.rgw.meta, "
+            "osd": "profile rbd-read-only, "
```
For upgrade cases, you need to update the

client.healthchecker in the list https://github.com/rook/rook/issues/12931#issuecomment-1729865450
And also need to test the upgrade scenario.

subhamkrai replied:
@parth-gr I added the changes, but could you help me with the upgrade steps? also, upstream and downstream upgrades are different steps, right?

parth-gr replied:
Create the caps with older versions of script, 
For upgrading use newer version  of script,
run the script with `python3 sript.py--upgrade` flag,

And see the permissions are getting upgraded from lower version of script to newer version

parth-gr replied:
Downstream is pretty much same.

But try to check with upgrade you are able to use your functionality

parth-gr replied:
Also check the upgrade works, I remember somewhat we should only add new permission to the end of the caps, but we can try this too

subhamkrai replied:
@parth-gr does that matter? I do see unit test is failing but not sure why

subhamkrai replied:
yes, it is just unit test we need to fix

Madhu-1 commented:
> > @subhamkrai are you able to test the feature completely with this user? i mean even the `ceph osd blocklist` was successful?
> 
> > @subhamkrai are you able to test the feature completely with this user? i mean even the `ceph osd blocklist` was successful?
> 
> No, @Madhu-1 I didn't verify the OSD blocklist command. Also, why need to verify the ` ceph osd blocklist` command the operator is not running that command right?

Sorry yes that's correct, we are creating the NetworkFence CR and NF CR will use CSI secret correct.

Madhu-1 reviewed (commented):
LGTM

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -632,7 +632,8 @@ def __init__(self, arg_list=None):
         self.MIN_USER_CAP_PERMISSIONS = {
             "mgr": "allow command config",
```
For the unit test, you need to add the new caps, `_init_cmd_output_map`

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1793,6 +1794,7 @@ def upgrade_users_permissions(self):
             "client.csi-cephfs-provisioner",
             "client.csi-rbd-node",
             "client.csi-rbd-provisioner",
+            "client.healthchecker",
```
You need to update the     `def get_caps_and_entity(self, user_name):`
and return the client and default caps of health checker,

Similar to what we have for other users,

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -969,6 +970,16 @@ def get_rbd_node_caps_and_entity(self):
 
         return caps, entity
 
+    def get_healthchecker_caps_and_entity(self):
```
hey, these values should not be hardcoded, as with the new script the default caps would change

You should update `create_checkerKey` how  `create_cephCSIKeyring_user` runs, 
 it should first `check_user_exist` and then only create it.

subhamkrai commented:
Now upgrade is also working,
```
python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name=replicapool
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "a=10.100.34.254:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "e863ab7f-1dfa-455b-bfff-ee3b4a93918f", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQB/txFlzQWSDRAAGjWGySfh7zk+DDUP705uBw=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.244.120.74", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQAQoxFlXHzeKBAAmtlAqLCoekOOzs+k8uwLqg=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQAQoxFl3wFGGxAAdaGO4u0vM9qEuEArVepfxw=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "http://10.244.120.74:7000/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "replicapool", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]

sh-4.4$ python3 /etc/ceph/create-external-cluster-resources.py --upgrade                       
Updated user client.csi-cephfs-node successfully.
Updated user client.csi-cephfs-provisioner successfully.
Updated user client.csi-rbd-node successfully.
Updated user client.csi-rbd-provisioner successfully.
Updated user client.healthchecker successfully.

sh-4.4$ 
```

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -986,6 +997,10 @@ def get_caps_and_entity(self, user_name):
                 self._arg_parser.restricted_auth_permission = True
             return self.get_rbd_node_caps_and_entity()
+        if "client.healthchecker" in user_name:
+            if "client.healthchecker" != user_name:
```
we can skip this check because I believe we only have one operator even if we are connected with multiple clusters

subhamkrai replied:
okay 👍🏽

subhamkrai replied:
I think, we can keep the check, if remove the caps, and the method was called to get the csi user caps and entity then also healthchecker user caps and entity will be retured

parth-gr replied:
I was talking about the `if "client.healthchecker" != user_name:`

subhamkrai replied:
removed

parth-gr replied:
I guess the changes are not commited, can you re-look

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -969,6 +970,16 @@ def get_rbd_node_caps_and_entity(self):
+        caps = {
+            "mgr": "allow command config",
+            "mon": "allow r, allow command quorum_status, allow command version",
+            "osd": "profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log,allow x pool=default.rgw.buckets.index",
```
Rgw pool prefix is not always same it can be set by the flag,
                    `self._arg_parser.rgw_pool_prefix`

subhamkrai replied:
can you suggest flag you mean to update?

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -969,6 +970,16 @@ def get_rbd_node_caps_and_entity(self):
+    def get_healthchecker_caps_and_entity(self):
+        entity = "client.healthchecker"
+        caps = {
+            "mgr": "allow command config",
```
this sequence is changed that's why unit test are not able to wok,

`mon` cap was above before

subhamkrai replied:
okay, I have updated but it's weird to fail test on that

parth-gr replied:
Actually the values above are hardcoded so we expect the values in the same order

parth-gr commented:
You need to add a command with 

`auth caps`, means a new self.cmd_output_map[] with this cmd
KeyError: '{"caps": ["mon", "allow r, allow command quorum_status, allow command version", "mgr", "allow command config", "osd", "profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index, allow rwx pool=.rgw.meta, allow rw pool=.rgw.control, allow rx pool=.rgw.log, allow x pool=.rgw.buckets.index"], "entity": "client.healthchecker", "format": "json", "prefix": "auth caps"}'

KeyError: '{"caps": ["mon", "allow r, allow command quorum_status, allow command version", "mgr", "allow command config", "osd", "profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log,allow x pool=default.rgw.buckets.index"], "entity": "client.healthchecker", "format": "json", "prefix": "auth get-or-create"}'

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -632,7 +635,8 @@ def __init__(self, arg_list=None):
         self.MIN_USER_CAP_PERMISSIONS = {
```
is this been used anywhere

subhamkrai replied:
this is inside_init_ method so the method is called not sure about that var

subhamkrai replied:
at least I don't the uses

parth-gr replied:
It was using before by the `create_checkerKey` but by the code changes we can remove this

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -969,6 +973,16 @@ def get_rbd_node_caps_and_entity(self):
+        caps = {
+            "mon": "allow r, allow command quorum_status, allow command version",
+            "mgr": "allow command config",
+            "osd": f"profile rbd-read-only, allow rwx pool={self._arg_parser.rgw_pool_prefix}.rgw.meta, allow r pool=.rgw.root, allow rw pool={self._arg_parser.rgw_pool_prefix}.rgw.control, allow rx pool={self._arg_parser.rgw_pool_prefix}.rgw.log,allow x pool={self._arg_parser.rgw_pool_prefix}.rgw.buckets.index",
```
Set this value in unit test `{self._arg_parser.rgw_pool_prefix}`

parth-gr replied:
It is an optional flag, so add an if on top of it either the flag value or the default value 
`DEFAULT_RGW_POOL_PREFIX`

parth-gr replied:
Can we check fir if else here `{self._arg_parser.rgw_pool_prefix}`

subhamkrai replied:
actually we're already setting the value but it was nor applying for upgrade so added for that also

parth-gr commented on deploy/examples/create-external-cluster-resources-tests.py:
```diff
@@ -84,8 +84,9 @@ def test_method_create_cephCSIKeyring_cephFSProvisioner(self):
     def test_non_zero_return_and_error(self):
         self.rjObj.cluster.return_val = 1
         self.rjObj.cluster.err_message = "Dummy Error"
+        self.rjObj.rgw_pool_prefix = "default"
```
set this in the upgrade test test_upgrade_user_permissions

subhamkrai replied:
added

parth-gr commented on deploy/examples/create-external-cluster-resources-tests.py:
```diff
@@ -175,6 +176,7 @@ def test_upgrade_user_permissions(self):
         )
         # for testing, we are using 'DummyRados' object
         self.rjObj.cluster = ext.DummyRados.Rados()
+        self.rjObj.rgw_pool_prefix = "default"
```
Arg parser is missing
self.rjObj._arg_parser.rgw_pool_prefix

subhamkrai commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -114,8 +114,8 @@ def _init_cmd_output_map(self):
-            """{"caps": ["mon", "allow r, allow command quorum_status", "osd", "allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow x pool=default.rgw.buckets.index"], "entity": "client.healthchecker", "format": "json", "prefix": "auth get-or-create"}"""
-        ] = """[{"entity":"client.healthchecker","key":"AQDFkbNeft5bFRAATndLNUSEKruozxiZi3lrdA==","caps":{"mon":"allow r, allow command quorum_status","osd":"allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow x pool=default.rgw.buckets.index"}}]"""
+            """{"caps": ["mon", "allow r, allow command quorum_status", "osd", "profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow x pool=default.rgw.buckets.index"], "entity": "client.healthchecker", "format": "json", "prefix": "auth get-or-create"}"""
+        ] = """[{"entity":"client.healthchecker","key":"AQDFkbNeft5bFRAATndLNUSEKruozxiZi3lrdA==","caps":{"mon":"allow r, allow command quorum_status","osd":"profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow x pool=default.rgw.buckets.index"}}]"""
```
The above is used for unit tests I guess for no impact on real code

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -632,7 +635,8 @@ def __init__(self, arg_list=None):
         self.MIN_USER_CAP_PERMISSIONS = {
```
remove as not used

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1467,7 +1484,9 @@ def _gen_output_map(self):
         self.out_map["ROOK_EXTERNAL_USERNAME"] = self.run_as_user
         self.out_map["ROOK_EXTERNAL_CEPH_MON_DATA"] = self.get_ceph_external_mon_data()
-        self.out_map["ROOK_EXTERNAL_USER_SECRET"] = self.create_checkerKey()
+        self.out_map["ROOK_EXTERNAL_USER_SECRET"] = self.create_checkerKey(
```
remove  the and not self._arg_parser.upgrade from the if else
`if not self._arg_parser.rgw_pool_prefix and not self._arg_parser.upgrade:`

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -145,7 +148,7 @@ def _init_cmd_output_map(self):
         self.cmd_output_map[
             """{"entity": "client.healthchecker", "format": "json", "prefix": "auth get"}"""
-        ] = """[{"entity":"client.healthchecker","key":"AQDFkbNeft5bFRAATndLNUSEKruozxiZi3lrdA==","caps":{"mon": "allow r, allow command quorum_status, allow command version", "mgr": "allow command config", "osd": "allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index"}}]"""
+        ] = """[{"entity":"client.healthchecker","key":"AQDFkbNeft5bFRAATndLNUSEKruozxiZi3lrdA==","caps":{"mon": "allow r, allow command quorum_status, allow command version", "mgr": "allow command config", "osd": "profile rbd-read-only, allow rwx pool=default.rgw.meta, allow r pool=.rgw.root, allow rw pool=default.rgw.control, allow rx pool=default.rgw.log, allow x pool=default.rgw.buckets.index ...
```
Upgrade help needs to modified, as we would be upgrading client.healthcker too and also we need to pass --rgw-pool-prefix flag if it was passed during the creation

 help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied."
            + "Sample run: `python3 /etc/ceph/create-external-cluster-resources.py --upgrade`, this will upgrade all the default csi users(non-restricted)"

parth-gr replied:
Also downstream docs need to modified

subhamkrai replied:
can you explaing more, somehow I'm not sure what you asking to change at this place?

subhamkrai replied:
Yes. this is resolved

subhamkrai commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -652,6 +644,21 @@ def __init__(self, arg_list=None):
             self.cluster = rados.Rados()
             self.cluster.conf_read_file()
         self.cluster.connect()
+        if not self._arg_parser.rgw_pool_prefix and not self._arg_parser.upgrade:
+            self._arg_parser.rgw_pool_prefix = self.DEFAULT_RGW_POOL_PREFIX
+        # when user has not provided rgw pool name during upgrade, get the rgw pool name used during connection
+        if not self._arg_parser.rgw_pool_prefix and self._arg_parser.upgrade:
+            _, val = self.check_user_exist("client.healthchecker")
+            # To get value 'default' which is rgw pool name from 'allow rwx pool=default.rgw.meta'
+            pattern = r"pool=(.*?)\.rgw\.meta"
+            match = re.search(pattern, val)
+            if match:
+                self._arg_parser.rgw_pool_prefix = match.group(1)
+                print(self._arg_parser.rgw_pool_prefix)
+            else:
+                raise ExecutionFailureException(
+                    "failed to get rgw pool name for upgrade"
+                )
```
```suggestion
        if not self._arg_parser.rgw_pool_prefix and not self._arg_parser.upgrade:
            self._arg_parser.rgw_pool_prefix = self.DEFAULT_RGW_POOL_PREFIX
        # when user has not provided rgw pool name during upgrade, get the rgw pool name used during connection
        if not self._arg_parser.rgw_pool_prefix and self._arg_parser.upgrade:
            self.run_as_user = self.EXTERNAL_USER_NAME
            _, val = self.check_user_exist("client.healthchecker")
            # To get value 'default' which is rgw pool name from 'allow rwx pool=default.rgw.meta'
            pattern = r"pool=(.*?)\.rgw\.meta"
            match = re.search(pattern, val)
            if match:
                self._arg_parser.rgw_pool_prefix = match.group(1)
                print(self._arg_parser.rgw_pool_prefix)
            else:
                raise ExecutionFailureException(
                    "failed to get rgw pool name for upgrade"
                )
```

The above change is used to get rgq pool name used during connecting the clusters the first time. The above changes avoid any doc changes and make upgrade steps same. Otherwise, we had to ask user to pass the rgw pool name they used during cluster connection.

parth-gr commented on deploy/examples/create-external-cluster-resources-tests.py:
```diff
@@ -84,8 +84,9 @@ def test_method_create_cephCSIKeyring_cephFSProvisioner(self):
     def test_non_zero_return_and_error(self):
         self.rjObj.cluster.return_val = 1
         self.rjObj.cluster.err_message = "Dummy Error"
+        self.rjObj._arg_parser.rgw_pool_prefix = "default"
```
this is not needed for non zero test

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -652,6 +644,20 @@ def __init__(self, arg_list=None):
+            self._arg_parser.rgw_pool_prefix = self.DEFAULT_RGW_POOL_PREFIX
+        # when user has not provided rgw pool name during upgrade, get the rgw pool name used during connection
+        if not self._arg_parser.rgw_pool_prefix and self._arg_parser.upgrade:
+            _, val = self.check_user_exist("client.healthchecker")
```
can we move this check to `    def upgrade_user_permissions(self, user): ` where we already get the user caps

subhamkrai replied:
The reason I kept under `_init` method is this is the first method called by python and here we are updating the cli flag `self._arg_parser.rgw_pool_prefix` so kept in this method. Also, other flags are modified here only.

parth-gr replied:
Actually this is just for an upgrade and I feel like we should make this separate and maybe even a separate function for this

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -652,6 +644,20 @@ def __init__(self, arg_list=None):
         self.cluster.connect()
+        if not self._arg_parser.rgw_pool_prefix and not self._arg_parser.upgrade:
+            self._arg_parser.rgw_pool_prefix = self.DEFAULT_RGW_POOL_PREFIX
+        # when user has not provided rgw pool name during upgrade, get the rgw pool name used during connection
```
# get the rgw pool name from the client.healthchecker caps

subhamkrai replied:
update

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1156,6 +1157,12 @@ def create_checkerKey(self):
+        user_key = self.check_user_exist(entity)
+        if user_key != "":
+            return user_key, f"{entity.split('.', 1)[1]}"
+            # entity.split('.',1)[1] to rename entity(client.csi-rbd-node) as csi-rbd-node
```
no need for this comment IMO

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1807,6 +1817,19 @@ def upgrade_user_permissions(self, user):
+        if user == "client.healthchecker":
+            # when admin has not provided rgw pool name during upgrade,
+            # get the rgw pool name from client.healthchecker user which was used during connection
+            if not self._arg_parser.rgw_pool_prefix:
```
can we separate this code in a function, the original function is becoming a lot big

subhamkrai replied:
done

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1807,6 +1817,19 @@ def upgrade_user_permissions(self, user):
             return
         existing_caps = json_out[0]["caps"]
+        if user == "client.healthchecker":
+            # when admin has not provided rgw pool name during upgrade,
```
As we are not documenting it in the above --upgrade flag help,

So we should not document it.

subhamkrai replied:
I think it helps in understanding the code what we are doing, so keeping it here also

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -471,7 +474,7 @@ def gen_arg_parser(cls, args_to_parse=None):
             action="store_true",
             default=False,
-            help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied."
+            help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) and client.healthchecker ceph users with new permissions needed for the new cluster version and older permission will still be applied."
```
the same thing needs to be updated in the external-cluster crd document.

subhamkrai replied:
updated

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -471,7 +474,7 @@ def gen_arg_parser(cls, args_to_parse=None):
             action="store_true",
             default=False,
-            help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) with new permissions needed for the new cluster version and older permission will still be applied."
+            help="Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) and client.healthchecker ceph users with new permissions needed for the new cluster version and older permission will still be applied."
```
Can you also share the before and after upgrade output of the health checker caps

subhamkrai replied:
added

parth-gr reviewed (commented):
Some nits

subhamkrai commented:
before upgrade:
```yaml
 kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "a=10.99.155.85:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "2147f6f5-60cd-4046-87b1-c329dd7add74", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQCNMBRlrYXwNxAA22pTgEgBZXmIuMw8jiid/w=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.244.120.91", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQB2KxRlkORXCxAArNrPFs9cxC1qtye1oEVRsg=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQB1KxRlHPIdNxAAWCKOY8fRyEeyavSeZT9utg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "http://10.244.120.91:7000/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "replicapool", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]
```

with upgrade:
```yaml
srai@192 ~ (update-caps) $ kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --upgrade
Updated user client.csi-cephfs-node successfully.
Updated user client.csi-cephfs-provisioner successfully.
Updated user client.csi-rbd-node successfully.
Updated user client.csi-rbd-provisioner successfully.
Updated user client.healthchecker successfully.

```

parth-gr commented on deploy/examples/create-external-cluster-resources.py:
```diff
@@ -1793,12 +1801,28 @@ def upgrade_users_permissions(self):
         for user in users:
             self.upgrade_user_permissions(user)
 
+    def get_rgw_pool_name_during_upgrade(self,user, caps):
```
I see a formating issue `, caps`
Can you run the python black formatter to make the formatting consistent

subhamkrai replied:
okay

subhamkrai commented:
connecting from master branch
```
srai@192 ~ (update-caps) $ git checkout master
Switched to branch 'master'
~/go/src/github.com/rook
srai@192 ~ (master) $ kubectl rook-ceph ceph auth del client.healthchecker
Warning: rook version 'rook: v1.12.0-alpha.0.291.gef487ec1c' is running a pre-release version of Rook.

Info: running 'ceph' command with args: [auth del client.healthchecker]
~/go/src/github.com/rook
srai@192 ~ (master) $  kubectl -n rook-ceph cp deploy/examples/create-external-cluster-resources.py $toolbox:/etc/ceph
~/go/src/github.com/rook
srai@192 ~ (master) $ kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool --rgw-pool-prefix=my-store
[{"name": "rook-ceph-mon-endpoints", "kind": "ConfigMap", "data": {"data": "a=10.99.155.85:6789", "maxMonId": "0", "mapping": "{}"}}, {"name": "rook-ceph-mon", "kind": "Secret", "data": {"admin-secret": "admin-secret", "fsid": "2147f6f5-60cd-4046-87b1-c329dd7add74", "mon-secret": "mon-secret"}}, {"name": "rook-ceph-operator-creds", "kind": "Secret", "data": {"userID": "client.healthchecker", "userKey": "AQALNRRlALjdCxAAji+P74cQy67HH3wkOUchjg=="}}, {"name": "monitoring-endpoint", "kind": "CephCluster", "data": {"MonitoringEndpoint": "10.244.120.91", "MonitoringPort": "9283"}}, {"name": "rook-csi-rbd-node", "kind": "Secret", "data": {"userID": "csi-rbd-node", "userKey": "AQB2KxRlkORXCxAArNrPFs9cxC1qtye1oEVRsg=="}}, {"name": "rook-csi-rbd-provisioner", "kind": "Secret", "data": {"userID": "csi-rbd-provisioner", "userKey": "AQB1KxRlHPIdNxAAWCKOY8fRyEeyavSeZT9utg=="}}, {"name": "rook-ceph-dashboard-link", "kind": "Secret", "data": {"userID": "ceph-dashboard-link", "userKey": "http://10.244.120.91:7000/"}}, {"name": "ceph-rbd", "kind": "StorageClass", "data": {"pool": "replicapool", "csi.storage.k8s.io/provisioner-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/controller-expand-secret-name": "rook-csi-rbd-provisioner", "csi.storage.k8s.io/node-stage-secret-name": "rook-csi-rbd-node"}}]
```

moving to local branch to test upgrade case
```
~/go/src/github.com/rook
srai@192 ~ (master) $ git checkout update-caps
Switched to branch 'update-caps'
~/go/src/github.com/rook
srai@192 ~ (update-caps) $  kubectl -n rook-ceph cp deploy/examples/create-external-cluster-resources.py $toolbox:/etc/ceph
~/go/src/github.com/rook
srai@192 ~ (update-caps) $ kubectl -n rook-ceph exec $toolbox -- python3 /etc/ceph/create-external-cluster-resources.py --upgrade
Updated user client.csi-cephfs-node successfully.
Updated user client.csi-cephfs-provisioner successfully.
Updated user client.csi-rbd-node successfully.
Updated user client.csi-rbd-provisioner successfully.
Updated user client.healthchecker successfully.

~/go/src/github.com/rook
srai@192 ~ (update-caps) $ kubectl rook-ceph ceph auth get client.healthchecker
Warning: rook version 'rook: v1.12.0-alpha.0.291.gef487ec1c' is running a pre-release version of Rook.

Info: running 'ceph' command with args: [auth get client.healthchecker]
[client.healthchecker]
    key = AQALNRRlALjdCxAAji+P74cQy67HH3wkOUchjg==
    caps mgr = "allow command config"
    caps mon = "allow r, allow command quorum_status, allow command version"
    caps osd = "allow rwx pool=my-store.rgw.meta, allow r pool=.rgw.root, allow rw pool=my-store.rgw.control, allow rx pool=my-store.rgw.log, allow x pool=my-store.rgw.buckets.index, profile rbd-read-only"````

parth-gr commented:
LGTM!

subhamkrai commented on Documentation/CRDs/Cluster/external-cluster.md:
```diff
@@ -31,35 +31,35 @@ Run the python script [create-external-cluster-resources.py](https://github.com/
+* `--rgw-realm-name`: (optional) Provides the name of the rgw-realm
+* `--rgw-zone-name`: (optional) Provides the name of the rgw-zone
+* `--rgw-zonegroup-name`: (optional) Provides the name of the rgw-zone-group
+* `--upgrade`: "(optional) Upgrades the cephCSIKeyrings(For example: client.csi-cephfs-provisioner) and client.healthchecker ceph users with new permissions needed for the new cluster version and older permission will still be applied."
```
right, removed
